CondTools/Ecal/plugins/EcalPedestalHistory.h

```diff
@@ -41,12 +41,12 @@
 #include <fstream>
 #include <map>
 
-class EcalPedestalHistory : public edm::EDAnalyzer {
+class EcalPedestalHistory : public edm::one::EDAnalyzer<> {
 public:
   explicit EcalPedestalHistory(const edm::ParameterSet&);
   ~EcalPedestalHistory() override;
   void analyze(const edm::Event&, const edm::EventSetup&) override;
-  void beginRun(edm::Run const&, edm::EventSetup const&) override;
+  void beginRun(edm::Run const&, edm::EventSetup const&); // no need to override beginRun when inheriting from one::EDAnalyzer 
```

Comment:
As mentioned above, the compile error that you probably got about the `override` keyword not overriding is there to protect you. `edm::one::EDAnalyzer<>` does not have a `beginRun()` by default. If something needs to be done at every beginning of a run the you can make it a `edm::one::EDAnalyzer<edm::one::WatchRuns>`, which has `beginRun()` and also `endRun()` and should have the `override` keyword.

Reply:
> If something needs to be done at every beginning of a run

This module clearly has some functionality in the beginRun() method, but I don't understand why we would want to switch to the `edm::one::EDAnalyzer<edm::one::WatchRuns>` module rather than just use the beginRun() definition included in this class.

Reply:
Again the same reason as for the other ones. The base class of `edm::one::EDAnalyzer<>` does not have a `beginRun()` and if you create ` beginRun()` in this class then the framework does not know about it and will never call it when it calls the `beginRun()` functions. It just happens that you have a function in `EcalPedestalHistory` that is called `beginRun()` as well but for the framework this does not matter since it is not in the base class. `edm::one::EDAnalyzer<edm::one::WatchRuns>` on the other hand does introduce a `beginRun()` and if your `EcalPedestalHistory` inherits from it the framework will call its `beingRun()`.

The legacy modules all had `beginRun()`, `beginLuminositySection()`, `beginJob()`, etc. defined all the time but the thread safe ones do not.

Reply:
So if you use `edm::one::EDAnalyzer<edm::one::WatchRuns>` you should be able to leave the `beginRun()` here as it was for the legacy module (including the `override`).

Reply:
Ok I see - so it's a matter of the "framework" knowing about the method, and this is only true if it comes from the new thread-safe classes which we are inheriting from.

Reply:
Yes. The framework (`cmsRun`) runs only the functions (`produce`, `analyze`, `beginRun`, `endJob`, etc.) of modules that it knows about. But not all modules have all functions (E.g. an EDProducer has no `analyze`). So the framework run the functions that are available in the base class of a module. The thread-safe module types also to not have all functions but can obtain some by template modifiers (like adding `beginRun` with the `edm::one::WatchRuns` template modifier).